### PR TITLE
Bug 1164266 - The way we update the UI job tables is slow. r=wlach

### PR DESCRIPTION
**Pull request by martiansideofthemoon**
@wlach ,
Well this is what I could make out from your patch. It doesn't exactly work, probably because I've messed up the HTML, though I can see the job symbols and click on the greyed out text to open the jobs panel. But in general I found it a really difficult task concatenating strings. I've added some questions on bugzilla.

[Review on Reviewable](https://reviewable.io/reviews/mozilla/treeherder/1235)

### CODE REVIEW

File: ui/js/directives/treeherder/clonejobs.js [hunk 1]
```diff
@@ -526,23 +526,31 @@ treeherder.directive('thCloneJobs', [
 
             //Empty the job column before populating it
             jobTdEl.empty();
-
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: What we want to do here is create a temporary variable like `var jobTdHtml` and start concatenating text to it. We'll only use a jQuery selector to add it to the DOM when we're done.

File: ui/js/directives/treeherder/clonejobs.js [hunk 2]
```diff
@@ -526,23 +526,31 @@ treeherder.directive('thCloneJobs', [
...
+                              jgObj.name + "\" data-grkey=\"" + jgObj.grkey +
+                              "\"><span class=\"group-symbol\" " +
+                              "ignore-job-clear-on-click=\"\">" + jgObj.symbol +
+                              "</span><span class=\"group-content\"><span class=\"group-job-list\">" +
+                              "<button class=\"btn job-btn btn-xs btn-dkgray key43 filter-shown\" " +
+                              "data-jmkey=\"key43\" ignore-job-clear-on-click=\"\" title=\"" +
+                              jgObj.jobs[0].job_type_name + "\">" + jgObj.jobs[0].job_type_symbol +
+                              "</button> </span><span class=\"group-count-list\">" +
+                              "</span></span></span>";
+                    jobTdEl.append(jobText);
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: We want to concatenate the text onto a string here (the one I suggest above), instead of appending to an array of dom elements.

File: ui/js/directives/treeherder/clonejobs.js [hunk 3]
```diff
@@ -526,23 +526,31 @@ treeherder.directive('thCloneJobs', [
...
+                    var jobText = "<span class=\"disabled job-group\" title=\"" +
+                              jgObj.name + "\" data-grkey=\"" + jgObj.grkey +
+                              "\"><span class=\"group-symbol\" " +
+                              "ignore-job-clear-on-click=\"\">" + jgObj.symbol +
+                              "</span><span class=\"group-content\"><span class=\"group-job-list\">" +
+                              "<button class=\"btn job-btn btn-xs btn-dkgray key43 filter-shown\" " +
+                              "data-jmkey=\"key43\" ignore-job-clear-on-click=\"\" title=\"" +
+                              jgObj.jobs[0].job_type_name + "\">" + jgObj.jobs[0].job_type_symbol +
+                              "</button> </span><span class=\"group-count-list\">" +
+                              "</span></span></span>";
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: So you could have saved some quoting troubles by using the "'" character. :) But a better solution would be to just use the interpolator here. The only difference is that you won't use the jQuery selector to create DOM elements here.

Something like:
```
jobTdHtml += jobGroupInterpolator(jobGroups[i])
```

The only tricky part is that there is some internal content which we need to put inside that object that we're interpolating (we were inserting it into the DOM before via jQuery selectors, below). So you'll need to change `addJobBtnEls` (below) to something like `getJobBtnEls` and make it return a blob of HTML text, which we then insert into the jobGroupInterpolator.

Maybe it would be helpful to describe how interpolators are instantiated here. They're just methods that return a JS string, defined in the clonejobs file here:

https://github.com/mozilla/treeherder/blob/d91cd4ba31d23bc1d4bebae5e040218584bacc72/ui/js/directives/treeherder/clonejobs.js#L46

They are based on templates which we are currently leaving in index.html. For example the job group interpolator is here:

https://github.com/mozilla/treeherder/blob/d91cd4ba31d23bc1d4bebae5e040218584bacc72/ui/index.html#L196

**Comment by martiansideofthemoon** [comment]
@wlach , How could I run lint tests locally? I'm just flooding the PR with lint fixups

File: ui/js/directives/treeherder/clonejobs.js [hunk 4]
```diff
@@ -366,11 +376,32 @@ treeherder.directive('thCloneJobs', [
...
             platformGroup.find(".group-count-list").html(jobCountBtnArray);
         };
 
+        var getGroupJobsAndCounts = function(jgObj) {
+            var btnArrays = renderGroupJobsAndCounts(jgObj);
+            var jobBtnArray = btnArrays.jobBtnArray;
+            var jobCountBtnArray = btnArrays.jobCountBtnArray;
+            var output = {
+                "jobBtnHTML": jobBtnArray.length ? jobBtnArray[0][0].outerHTML : "",
+                "jobCountBtnHTML": jobCountBtnArray.length ? jobCountBtnArray[0][0].outerHTML : ""
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: There's no need for the quotes here. Something like:

```js
var output = {
  jobBtnHtml: ...,
  jobCountBtnHTML: ...
}
```

should work fine

File: ui/js/directives/treeherder/clonejobs.js [hunk 5]
```diff
@@ -366,11 +376,32 @@ treeherder.directive('thCloneJobs', [
...
 
+        var addGroupJobsAndCounts = function(jgObj, platformGroup) {
+            var btnArrays = renderGroupJobsAndCounts(jgObj);
+            var jobBtnArray = btnArrays.jobBtnArray;
+            var jobCountBtnArray = btnArrays.jobCountBtnArray;
             platformGroup.find(".group-job-list").html(jobBtnArray);
             platformGroup.find(".group-count-list").html(jobCountBtnArray);
         };
 
+        var getGroupJobsAndCounts = function(jgObj) {
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: Since this is only being called once, I'd just inline it where it's being called.

File: ui/js/directives/treeherder/clonejobs.js [hunk 6]
```diff
@@ -526,29 +557,29 @@ treeherder.directive('thCloneJobs', [
...
                 jgObj = jobGroups[i];
-
                 if (jgObj.symbol !== '?') {
                     // Job group detected, add job group symbols
                     jobGroups[i].grkey = jgObj.mapKey;
                     jobGroups[i].collapsed = true;
-                    jobGroup = $(jobGroupInterpolator(jobGroups[i]));
-                    jobTdEl.append(jobGroup);
-
+                    jobGroups[i].btnHTML = getJobBtnEls(jgObj);
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: I don't think line makes sense -- we reset the value immediately below.

File: ui/js/directives/treeherder/clonejobs.js [hunk 7]
```diff
@@ -526,29 +557,29 @@ treeherder.directive('thCloneJobs', [
...
                     if (isGroupExpanded(jgObj)) {
-                        addJobBtnEls(jgObj, jobGroup.find(".group-job-list"));
+                        jobGroups[i].btnHTML = addJobBtnEls(jgObj);
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: I don't think this will have the effect that we want, since addJobBtnEls will call renderJobBtnEls, which seems to do a bunch of DOM manipulation. So I'd be surprised if we're getting much benefit from things as they stand.

We need to rework this function so it doesn't call any jquery selectors (`$(...)`) at all while it's working. Only string manipulations.

File: ui/js/directives/treeherder/clonejobs.js (hunk 5 above)
Comment (martiansideofthemoon) [inline ui/js/directives/treeherder/clonejobs.js]: @wlach , Not so sure what you mean here. Could you give me an example or a link to some sort of guide / documentation? Thanks!

File: ui/js/directives/treeherder/clonejobs.js (hunk 6 above)
Comment (martiansideofthemoon) [inline ui/js/directives/treeherder/clonejobs.js]: Yeah that was old code to be honest, before I wrote the `if`.

File: ui/js/directives/treeherder/clonejobs.js (hunk 5 above)
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: I mean just remove the function and put the contents where it's called. :)

File: ui/js/directives/treeherder/clonejobs.js [hunk 8]
```diff
@@ -526,29 +566,29 @@ treeherder.directive('thCloneJobs', [
...
-                        addJobBtnEls(jgObj, jobGroup.find(".group-job-list"));
+                        jobGroups[i].btnHTML = getJobBtnEls(jgObj);
                     } else {
-                        addGroupJobsAndCounts(jgObj, jobGroup);
+                        var html = getGroupJobsAndCounts(jgObj);
+                        jobGroups[i].btnHTML = html.jobBtnHTML;
+                        jobGroups[i].countBtnHTML = html.jobCountBtnHTML;
                     }
+                    jobTdHtml += jobGroupInterpolator(jobGroups[i]);
+                    console.log(jobTdHtml);
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: This console.log should be removed.

File: ui/js/directives/treeherder/clonejobs.js [hunk 9]
```diff
@@ -360,17 +371,46 @@ treeherder.directive('thCloneJobs', [
...
+            var jobCountBtnArray = btnArrays.jobCountBtnArray;
             platformGroup.find(".group-job-list").html(jobBtnArray);
             platformGroup.find(".group-count-list").html(jobCountBtnArray);
         };
 
+        var getGroupJobsAndCounts = function(jgObj) {
+            var btnArrays = renderGroupJobsAndCounts(jgObj);
+            var jobBtnArray = btnArrays.jobBtnArray;
+            var jobCountBtnArray = btnArrays.jobCountBtnArray;
+            var jobBtnHTML = "";
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: Just a general point, I don't think it's a good idea to have so many temporary variables here. Unless the expression you're evaluating is really long or is used more than once or twice, it's better just to use it directly. i.e. you could have just done the forEach against btnArrays.jobCountBtnArray directly.

File: ui/js/directives/treeherder/clonejobs.js [hunk 10]
```diff
@@ -526,29 +566,29 @@ treeherder.directive('thCloneJobs', [
 
             //Empty the job column before populating it
             jobTdEl.empty();
-
-            var jgObj, jobGroup, i;
+            var jgObj, jobGroup, i, jobTdHtml = "";
             for (i=0; i<jobGroups.length; i++) {
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: Since we're modifying so much here anyway, it would be nice to just use a `forEach` here, that would let us get rid of the `i` temporary.

File: ui/js/directives/treeherder/clonejobs.js [hunk 11]
```diff
@@ -360,17 +371,46 @@ treeherder.directive('thCloneJobs', [
                     countInfo.value = countInfo.count;
                     countInfo.title = countInfo.count + " " + countInfo.countText + " jobs in group";
                     countInfo.btnClass = countInfo.btnClass + "-count";
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: Could we not perform this logic inside the interpolator, instead of assigning to a temporary? Seems to me like that would be more clear.

File: ui/js/directives/treeherder/clonejobs.js [hunk 12]
```diff
@@ -526,29 +566,29 @@ treeherder.directive('thCloneJobs', [
...
                 if (jgObj.symbol !== '?') {
                     // Job group detected, add job group symbols
                     jobGroups[i].grkey = jgObj.mapKey;
                     jobGroups[i].collapsed = true;
-                    jobGroup = $(jobGroupInterpolator(jobGroups[i]));
-                    jobTdEl.append(jobGroup);
-
                     if (isGroupExpanded(jgObj)) {
-                        addJobBtnEls(jgObj, jobGroup.find(".group-job-list"));
+                        jobGroups[i].btnHTML = getJobBtnEls(jgObj);
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: Instead of assigning to temporaries and then interpolating, I'd just call the interpolator directly.
i.e.
```js
jobTdHtml += jobGroupInterpolator({ btnHTML: getJobBtnEls(jgObj), ... })
```

Ditto below.

Comment (martiansideofthemoon) [inline ui/js/directives/treeherder/clonejobs.js]: Didn't do this one, I don't think it applies any more as I've moved the function to this place

File: ui/js/directives/treeherder/clonejobs.js [hunk 13]
```diff
@@ -357,16 +368,28 @@ treeherder.directive('thCloneJobs', [
...
+                    });
                     jobCountBtnArray.push(jobCountBtn);
                     jobCountBtnArray.push(' ');
                 }
             });
+            var output = {
+                jobBtnArray: jobBtnArray,
+                jobCountBtnArray: jobCountBtnArray
+            };
+            return output;
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: Let's just return `{ jobBtnArray: ..., ... }` directly instead of assigning it to a temporary.

File: ui/js/directives/treeherder/clonejobs.js [hunk 14]
```diff
@@ -357,16 +368,28 @@ treeherder.directive('thCloneJobs', [
...
-                    var jobCountBtn = $(jobGroupCountInterpolator(countInfo));
-                    showHideElement(jobCountBtn, true);
+                    var jobCountBtn = jobGroupCountInterpolator({
+                        value: countInfo.count,
+                        title: countInfo.count + " " + countInfo.countText + " jobs in group",
+                        btnClass: countInfo.btnClass + "-count",
+                        visibleClass: "filter-shown",
+                        selectedClasses: countInfo.selectedClasses
+                    });
                     jobCountBtnArray.push(jobCountBtn);
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: Let's add the value of jobGroupCountInterpolator to the array directly, instead of assigning it to a temporary first.

File: ui/js/directives/treeherder/clonejobs.js [hunk 15]
```diff
@@ -526,29 +549,33 @@ treeherder.directive('thCloneJobs', [
 
             //Empty the job column before populating it
             jobTdEl.empty();
-
-            var jgObj, jobGroup, i;
-            for (i=0; i<jobGroups.length; i++) {
-
-                jgObj = jobGroups[i];
-
+            var jgObj, jobTdHtml = "";
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: If you look carefully, you'll see that the jgObj temporary is no longer necessary, and we can just replace all instances of it with `jobGroup`.

File: ui/js/directives/treeherder/clonejobs.js [hunk 16]
```diff
@@ -526,29 +549,33 @@ treeherder.directive('thCloneJobs', [
...
+            jobGroups.forEach(function (jobGroup) {
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: there shouldn't be a space between `function` and `(jobGroup)`

File: ui/js/directives/treeherder/clonejobs.js [hunk 17]
```diff
@@ -526,29 +549,33 @@ treeherder.directive('thCloneJobs', [
...
-
+                    jobGroup.grkey = jgObj.mapKey;
+                    jobGroup.collapsed = true;
                     if (isGroupExpanded(jgObj)) {
-                        addJobBtnEls(jgObj, jobGroup.find(".group-job-list"));
+                        jobGroup.btnHTML = getJobBtnEls(jgObj);
                     } else {
-                        addGroupJobsAndCounts(jgObj, jobGroup);
+                        var btnArrays = renderGroupJobsAndCounts(jgObj);
+                        jobGroup.btnHTML = "";
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: I think it's better not to assign btnHTML to jobGroup, since that's going to stick around and we don't want to unnecessarily consume memory. Instead, let's just stick it and countBtnHTML in temporary variables, then create an object to pass to jobGroupInterpolator below:

```js
var btnHTML, countBtnHTML;
...
jobTdHtml += jobGroupInterpolator({
    btnHTML: btnHTML,
    countBtnHTML: countBtnHTML,
    ...
})
```

File: ui/js/directives/treeherder/clonejobs.js [hunk 18]
```diff
@@ -526,29 +549,33 @@ treeherder.directive('thCloneJobs', [
...
                     }
+                    jobTdHtml += jobGroupInterpolator(jobGroup);
                 } else {
                     // Add the job btn spans
-                    addJobBtnEls(jgObj, jobTdEl);
+                    jobTdHtml += getJobBtnEls(jgObj);
                 }
-            }
+            });
+            jobTdEl.append($(jobTdHtml));
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: I think we can replace this with `jobTdEl.html(jobTdHtml)`: http://api.jquery.com/html/#html2

That would make the `jobTdEl.empty()`, above, unnecessary.

File: ui/js/directives/treeherder/clonejobs.js [hunk 19]
```diff
@@ -239,9 +239,22 @@ treeherder.directive('thCloneJobs', [
 
                 addJobBtnToArray(job, lastJobSelected, jobBtnArray);
             }
+            return jobBtnArray;
+        };
+        var addJobBtnEls = function(jgObj, jobList) {
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: I think it might be better to just remove this function and inline the code directly in the two places that it's called, since it's so simple.

**Comment by wlach** [comment]
We don't actually use the r= syntax for Treeherder. Also, generally it's better for a commit message to describe what was done, rather than the reason. :) Could you change the commit message to something like `Bug 1164266 - Improve the performance of job table rendering`?

**Comment by martiansideofthemoon** [comment]
Oh alright, I just copied the text from Bugzilla :stuck_out_tongue_closed_eyes: Do we squash commits in Treeherder?

File: ui/js/directives/treeherder/clonejobs.js [hunk 20]
```diff
@@ -249,18 +262,17 @@ treeherder.directive('thCloneJobs', [
...
-                    jobBtn.addClass("runnable-job-btn-selected");
+                    jobStatus.runnableClass = "runnable-job-btn-selected";
                 }
+                jobBtn = runnableJobBtnInterpolator(jobStatus);
             } else {
-                jobBtn = $(jobBtnInterpolator(jobStatus));
+                jobBtn = jobBtnInterpolator(jobStatus);
             }
 
             // If the job is currently selected make sure to re-apply
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: Just realized that this block of code needs to be updated to handle the fact that we no longer have a jQuery element to add the class to. You'll need to add the classes applied in `setSelectJobStyles` to the interpolated job element, then re-select the job once it exists. The first part should be relatively straightforward, the second might be tricky -- probably the right way of doing this is to make the resultset store model no longer keep a reference to the job element -- we should get that element ourselves where needed based purely on the id. 

If you want to try tackling this, you'll need to modify `ui/js/models/resultsets_store.js`, in particular the getSelectedJob and setSelectedJob to only keep a reference to the selected job's id, not el. You'll also need to add a jQuery selector inside clonejobs.js to get the element dynamically based on the id (where we were getting it via getSelectedJob before).

Comment (martiansideofthemoon) [inline ui/js/directives/treeherder/clonejobs.js]: @wlach, I finished with all the other issues, just a bit of this one left. I can't seem to understand what to store in place of `el`. There is the job ID, that's anyway stored under `repositories[repoName].lastJobObjSelected.id`. So I can use this ID maybe and fetch the element dynamically when needed. How can I use this ID with jQuery? I don't think `$("#" + job.id)` will work as this doesn't seem to be an HTML id

File: ui/js/directives/treeherder/clonejobs.js [hunk 21]
```diff
@@ -219,8 +222,11 @@ treeherder.directive('thCloneJobs', [
             return visible;
         };
 
-        var addJobBtnEls = function(jgObj, jobList) {
+        var renderJobBtnEls = function(jgObj) {
             var lastJobSelected = ThResultSetStore.getSelectedJob($rootScope.repoName);
+            if (!_.isEmpty(lastJobSelected.job)) {
+                lastJobSelected.el = $('.' + getJobMapKey(lastJobSelected.job));
+            }
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: I *think* this will work, but it should be called after we've rerendered all the job rows, not every time we render them, perhaps with a comment like:
```js
// update the selected job element, in case it changed
```
I would put this at the end of the two functions which wind up calling this: generateJobElements and updateJobs.

File: ui/js/models/resultsets_store.js [hunk 22]
```diff
@@ -255,14 +255,12 @@ treeherder.factory('ThResultSetStore', [
         };
 
         var getSelectedJob = function(repoName){
-            return { el:repositories[repoName].lastJobElSelected,
```
Comment (wlach) [inline ui/js/models/resultsets_store.js]: If you're going to take the approach above (of getting it via a selector after updating the job tables), you should revert this change.

File: ui/js/models/resultsets_store.js [hunk 23]
```diff
@@ -255,14 +255,12 @@ treeherder.factory('ThResultSetStore', [
...
+            return { el:{},
                      job:repositories[repoName].lastJobObjSelected };
         };
 
         var setSelectedJob = function(
             repoName, lastJobElSelected, lastJobObjSelected){
-
-            repositories[repoName].lastJobElSelected = lastJobElSelected;
```
Comment (wlach) [inline ui/js/models/resultsets_store.js]: Likewise, revert this change if you're going to update this element via a selector.

File: ui/js/directives/treeherder/clonejobs.js (hunk 21 above)
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: Note that calling jQuery functions isn't inherently expensive/problematic, it's calling them 100, 1000, 10000, 100000 times when they become a big issue. :)

File: ui/js/directives/treeherder/clonejobs.js [hunk 24]
```diff
@@ -249,32 +256,32 @@ treeherder.directive('thCloneJobs', [
...
-
             // If the job is currently selected make sure to re-apply
             // the job selection styles
             if( !_.isEmpty(lastJobSelected.job) &&
                 (lastJobSelected.job.id === job.id)){
-
-                setSelectJobStyles(jobBtn);
-
                 //Update the selected job element to the current one
                 ThResultSetStore.setSelectedJob($rootScope.repoName, jobBtn, job);
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: Is this line necessary? It seems to me like the call to selectJob at the end will fix this up.

File: ui/js/directives/treeherder/clonejobs.js [hunk 25]
```diff
@@ -249,32 +256,32 @@ treeherder.directive('thCloneJobs', [
...
-                jobBtn = $(jobBtnInterpolator(jobStatus));
+                jobBtn = jobBtnInterpolator(jobStatus);
             }
-
             // If the job is currently selected make sure to re-apply
             // the job selection styles
             if( !_.isEmpty(lastJobSelected.job) &&
                 (lastJobSelected.job.id === job.id)){
-
-                setSelectJobStyles(jobBtn);
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: I think we need something like this, otherwise there will be an interval where the selected job won't have the correct styles applied.

File: ui/js/directives/treeherder/clonejobs.js [hunk 26]
```diff
@@ -127,7 +127,6 @@ treeherder.directive('thCloneJobs', [
         var selectJob = function(job, job_selection_type) {
             var jobKey = getJobMapKey(job);
             var jobEl = $('.' + jobKey);
-
             clickJobCb({}, jobEl, job, job_selection_type);
             scrollToElement(jobEl);
```
Comment (wlach) [inline ui/js/directives/treeherder/clonejobs.js]: I don't think we want to emit a signal that the selected job has changed if we're just re-adding the selecting job styles after an update. Likewise I don't think we want to re-scroll to the selected job.

File: ui/js/directives/treeherder/clonejobs.js (hunk 25 above)
Comment (martiansideofthemoon) [inline ui/js/directives/treeherder/clonejobs.js]: @wlach, how can we set the styles when the element has not been created? we are just concatenating HTML at this stage right?

File: ui/js/directives/treeherder/clonejobs.js (hunk 26 above)
Comment (martiansideofthemoon) [inline ui/js/directives/treeherder/clonejobs.js]: @wlach , yeah you are right, I was putting the wrong object in the jQuery selector

File: ui/js/directives/treeherder/clonejobs.js (hunk 24 above)
Comment (martiansideofthemoon) [inline ui/js/directives/treeherder/clonejobs.js]: no it isn't. removing it!

**Comment by wlach** [comment]
This looks good! I'm going to squash this patch and land it.

**Comment by wlach** [comment]
Merged in 74dc50e0cb9eb9771ef8dc4a6e710e934e914e9c
